api/outdated/oidc_auth/models.py

```diff
@@ -0,0 +1,67 @@
+import logging
+
+from django.conf import settings
+
+from outdated.user.models import User
+
+logger = logging.getLogger(__name__)
+
+
+class BaseUser:  # pragma: no cover
+    def __init__(self):
+        self.email = None
+        self.groups = []
+        self.group = None
+        self.token = None
+        self.claims = {}
+        self.is_authenticated = False
+
+    def __str__(self):
+        raise NotImplementedError
+
+
+class OIDCUser(BaseUser):
+    def __init__(self, token: str, claims: dict):
+        super().__init__()
+
+        self.claims = claims
+        self.id = self.claims[settings.OIDC_CLAIMS["ID"]]
+        self.email = self.claims.get(settings.OIDC_CLAIMS["EMAIL"])
+        self.groups = self.claims.get(settings.OIDC_CLAIMS["GROUPS"], [])
+        self.group = self.groups[0] if self.groups else None
+        self.first_name = self.claims.get(settings.OIDC_CLAIMS["FIRST_NAME"])
+        self.last_name = self.claims.get(settings.OIDC_CLAIMS["LAST_NAME"])
+        self.username = self.claims.get(settings.OIDC_CLAIMS["USERNAME"])
+
+        self.token = token
+        self.is_authenticated = True
+        self.user = self._get_or_create_user()
+
+    def _get_or_create_user(self):
+        try:
+            user = User.objects.get(idp_id=self.id)
+
+            if (
+                user.email != self.email
+                or user.first_name != self.first_name
+                or user.last_name != self.last_name
+                or user.username != self.username
+            ):
+                user.email = self.email
+                user.first_name = self.first_name
+                user.last_name = self.last_name
+                user.username = self.username
+                user.save()
+
+        except User.DoesNotExist:
+            user = User.objects.create(
+                idp_id=self.id,
+                username=self.username,
+                email=self.email,
+                first_name=self.first_name,
+                last_name=self.last_name,
+            )
```

Comment:
Untested, but should work:tm: 

```suggestion
        user, _ = User.objects.update_or_create(
            idp_id=self.id,
            defaults={
                "email": self.email,
                "first_name": self.first_name,
                "last_name": self.last_name,
                "username": self.username,
            },
        )
```